articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+
+As mentioned earlier, malicious users may double encode their unicode input by additionally encoding the backslash, `\`, and `u`,
+which allows that to be resolved first in your client application code, and then resolved as the actual attack in the database.
+For example; `\u005C\u00750027` may be sent to the database as the input `\u0027`, thus successfully escaping the single quotation.
+
+When it comes to the validation of input such as usernames, it is often done to avoid the use of reserved keywords, such as admin.
+Unicode encoding can be used as another common bypass for this.
+For example the user input `\u0061\u0064\u006d\u0069\u006e` is the unicode for `admin`:
 
-Consider this query:
 
 [source,cypher]
 ----
-CALL apoc.cypher.doIt("CREATE (s:Student) SET s.name = '" + $studentName + "' RETURN true", {}) YIELD value
-RETURN value;
+CREATE (n {username: '\u0061\u0064\u006d\u0069\u006e'}) RETURN n.username
 ----
 
-Even though we passed `$studentName` as a parameter to the query, the parameter is being appended into the query string for execution by `apoc.cypher.doIt()`.
-The same vulnerability would exist if we were processing a LOAD CSV query like before, and appending into a query string for execution by APOC.
+.Results
+[opts="header"]
+|===
+|n.username
+|"Admin"
+|===
+
+=== String Concatenation
+
+Another method to bypass validation of particular keywords is using string concatenation in the injection.
+For example, validation that the user isn’t setting their username to admin could be bypassed with the injection:
+
+`"ad' + 'min'"`
+
+=== Second Order Injections
+
+A second order injection occurs when the input is successfully filtered and sanitized the first time it is used and is
+then stored in the database, when the application uses the value another time, the malicious code is then executed.
 
-Little Robby Labels would end up wiping out all our student data.
+For example; Little Robby Labels sets up an account with a username and password:
 
-== Pass parameters to APOC procs instead of appending into the query string
+`Username: LilRob' OR 1=1 WITH true AS hacked MATCH (a) DETACH DELETE ALL; –`
 
-Parameter usage here is still the answer to this vulnerability.
+As the username is received from the user directly, our application runs sanitization on it, escaping the quotation mark.
 
-Instead of appending into the string, pass the value as a parameter to the procedure call:
 
 [source,cypher]
 ----
-CALL apoc.cypher.doIt("CREATE (s:Student) SET s.name = $name  RETURN true", {name:$name}) YIELD value
-RETURN value;
+CREATE (u:User) SET u.username = 'LilRob\' OR 1=1 WITH true AS hacked MATCH (a) DETACH DELETE ALL; –';
 ----
 
-Little Robby Labels becomes powerless here.
+Little Robby Tables now resets their password and as the username is assumed to be safe our application runs sanitization only the new password:
+
```

Comment:
I think in these steps we kind of hop too quickly to the conclusion without really explaining that the username is retrieved from the database first.